pythia/models/butd.py

```diff
@@ -108,52 +109,92 @@ def get_data_t(self, t, data, batch_size_t, prev_output):
         output = None
         batch_size_t = batch_size
         for t in range(timesteps):
             data, batch_size_t = self.get_data_t(t, data, batch_size_t, output)
-            if self.config["inference"]["type"] == "beam_search":
-                pi_t = data["texts"]
-            else:
-                pi_t = data["texts"][:, t].unsqueeze(-1)
+            pi_t = data["texts"]
             embedding = self.word_embedding(pi_t)
             attention_feature, _ = self.process_feature_embedding(
                 "image", sample_list, embedding[:, 0, :], batch_size_t=batch_size_t
             )
             output = self.classifier(attention_feature)
-
             # Compute Beam Search decoding
-            if self.config["inference"]["type"] == "beam_search":
-                finish, data, batch_size_t = beam_search.search(t, data, output)
-                if finish:
-                    break
-            else:
-                scores[:batch_size_t, t] = output
+            finish, data, batch_size_t = beam_search.search(t, data, output)
+            if finish:
+                break
 
         model_output = {"scores": scores}
-        if self.config["inference"]["type"] == "beam_search":
-            model_output["captions"] = beam_search.best_score()
+        model_output["captions"] = beam_search.best_score()
 
         return model_output
+
+    def forward_nucleus_sampling(self, sample_list, scores):
+        nucleus_sampling = NucleusSampling(
+            self.vocab, self.config["inference"]["params"]["sum_threshold"]
+        )
+        sample_list = nucleus_sampling.init_batch(sample_list)
+        batch_size = sample_list.image_feature_0.size(0)
+        data, sample_list, timesteps = self.prepare_data(sample_list, batch_size)
+        output = None
+        batch_size_t = batch_size
+        for t in range(timesteps):
+            data, batch_size_t = self.get_data_t(t, data, batch_size_t, output)
+            pi_t = data["texts"]
+            embedding = self.word_embedding(pi_t)
+            attention_feature, _ = self.process_feature_embedding(
+                "image", sample_list, embedding[:, 0, :], batch_size_t=batch_size_t
+            )
+            output = self.classifier(attention_feature)
+            # Compute Nucleus Sampling decoding
+            finish, data, batch_size_t = nucleus_sampling.sample(t, data, output)
+            if finish:
+                break
+
+        model_output = {"scores": scores}
+        model_output["captions"] = nucleus_sampling.get_caption()
+
+        return model_output
+
+    def forward(self, sample_list):
+        # Stores the output probabilites.
+        scores = sample_list.answers.new_ones(
+            (
+                sample_list.answers.size(0),
+                self.text_processor.max_length,
+                self.vocab_size,
+            ),
+            dtype=torch.float,
+        )
+        if self.config["inference"]["type"] == "beam_search":
```

Comment:
maybe replace these if conditions with a `getattr()` So something like:
`forward_fn = getattr(self, 'forward_' + self.config["inference"]["type"])`

Reply:
Makes sense. Will commit the required changes.